pkg/build/build_test.go

```diff
@@ -96,72 +49,108 @@ func TestLoadConfiguration(t *testing.T) {
 		Members:   []string{"build"},
 	}}
 
-	tests := []struct {
-		description string
-		contents    string
-		template    string
-		expected    *Configuration
-		shouldErr   bool
-	}{
-		{
-			description: "no template",
-			contents:    defaultTemplateYaml,
-			expected:    expected,
-		}, {
-			description: "valid template",
-			contents:    templatized,
-			template:    `{"Package": "nginx", "Version": 100}`,
-			expected:    expected,
-		}, {
-			description: "incomplete template",
-			contents:    templatized,
-			template:    `{"Package": "nginx"}`,
-			shouldErr:   true,
-		}, {
-			description: "invalid template",
-			contents:    templatized,
-			template:    `{"Hello": "world"}`,
-			shouldErr:   true,
-		},
+	f := filepath.Join(t.TempDir(), "config")
+	if err := os.WriteFile(f, []byte(contents), 0755); err != nil {
+		t.Fatal(err)
 	}
 
-	for _, test := range tests {
-		t.Run(test.description, func(t *testing.T) {
-			dir := t.TempDir()
-			f := filepath.Join(dir, "config")
-			if err := ioutil.WriteFile(f, []byte(test.contents), 0755); err != nil {
-				t.Fatal(err)
-			}
-
-			cfg := &Configuration{}
-			err := cfg.Load(f, test.template)
-			if test.shouldErr && err == nil {
-				t.Fatal("expected test to fail but it passed")
-			}
-			if test.shouldErr {
-				return
-			}
-			if d := cmp.Diff(cfg, test.expected); d != "" {
-				t.Fatalf("actual didn't match expected: %s", d)
-			}
-		})
+	cfg := &Configuration{}
+	if err := cfg.Load(f); err != nil {
+		t.Fatal(err)
+	}
+	if d := cmp.Diff(expected, cfg); d != "" {
+		t.Fatalf("actual didn't match expected: %s", d)
 	}
 }
 
-// Makes sure the substitution map in pipeline.go matches the one in build.go
-// TODO: priyawadhwa@, it would be better to use a single map so we don't risk losing substitutions
-func TestSubstitutionReplacementMap(t *testing.T) {
-	pipelineMap := substitutionMap(&PipelineContext{
-		Package:    &Package{Name: "package"},
-		Subpackage: &Subpackage{Name: "subpackage"},
-		Context:    &Context{},
-	})
-	buildMap := substitutionReplacements()
+func TestLoadConfiguration_RangeSubpackages(t *testing.T) {
+	contents := `
+package:
+  name: hello
+  version: world
+
+pipeline:
+- name: hello
+  runs: world
+
+data:
+  - name: ninja-turtles
+    items:
+    - key: Michelangelo
+      value: orange
+    - key: Raphael
+      value: red
+    - key: Leonardo
+      value: blue
+    - key: Donatello
+      value: purple
+  - name: animals
+    items:
+    - key: dogs
+      value: loyal
+    - key: cats
+      value: angry
+    - key: turtles
+      value: slow
```

Comment:
Did this end up being better/easier than a `map[string]string` approach? E.g. where the YAML would look like

```yaml
items:
  dogs: loyal
  cats: angry
  turtles: slow
```

Reply:
I think we can't use a map here, because the ordering won't be preserved, and @kaniini indicated that ordering might matter. Maybe that's only true between explicitly defined subpackages, and if it doesn't hold for matrix-generated subpackage defs then we can go back to maps for this, and just sort the generated subpackages.

I do think that `key: value` is much easier to read and write than `- key: key; value: value` on multiple lines, so it'd be great if we could get back to a map.

Reply:
We might be able to implement a custom unmarshaler for this part of the data structure. Where for people writing YAML it looks like it'd be `map[string[string`, but really on the Go side, it's something else that preserves order (like `[]struct{key, value string}` or something).

But also, I think what's already there is totally good enough, too 👍

Reply:
i would rather do the unmarshaller myself.

Reply:
This is done now. The result is that the order specified in the YAML map is not kept, and items are just sorted by key for stability, which matches how maps are YAML-marshaled anyway.